source/server/http/admin.cc

```diff
@@ -763,7 +763,14 @@ Http::Code AdminImpl::handlerDrainListeners(absl::string_view url, Http::Respons
   ListenerManager::StopListenersType stop_listeners_type =
       params.find("inboundonly") != params.end() ? ListenerManager::StopListenersType::InboundOnly
                                                  : ListenerManager::StopListenersType::All;
-  server_.listenerManager().stopListeners(stop_listeners_type);
+  if (params.find("graceful") != params.end()) {
```

Comment:
1/2. From a release/docs perspective, how annoying would it be to change `?graceful` from a `bool -> int`? We're likely going to require in the future that the default drain timeout be overrideable via a query parameter. I was hoping to do that work in a separate PR, but if the type change release process is a hassle I'll see if I can get it to work now.

3. I have no opinion on this, I'm not really familiar with the InboundOnly use case. But actually this would avoid the (immediate) need for a query parameter and resolve the #1/2 problem, so I'd be ok with removing the query param, and just graceful draining on StopListenersType::All for now.

4. No, our use case needs those steps to be separate. There's a first phase with connection-discouraging behaviour, *then* listeners are closed and new connections are not accepted (but in-flight work can be completed), *then* all connections are terminated. We allow the first phase time period to be configured to 0 though, which would cause the behaviour you describe here.

Reply:
re: 1/2, my advice would be to use different query parameters for different things. So add `graceful`, then later add `drain_timeout` or whatever, etc. I think this will be more clear anyway.

re: 3 yeah sounds good.

re: 4, OK, let's just make sure it's clearly documented and tested.